turbopack/crates/turbo-tasks-fs/src/read_glob.rs

```diff
@@ -40,29 +42,85 @@ async fn read_glob_internal(
     let dir = directory.read_dir().await?;
     let mut result = ReadGlobResult::default();
     let glob_value = glob.await?;
+    let handle_file = |result: &mut ReadGlobResult,
+                       entry_path: &RcStr,
+                       segment: &RcStr,
+                       entry: &DirectoryEntry| {
+        if glob_value.matches(entry_path) {
+            result.results.insert(segment.clone(), entry.clone());
+        }
+    };
+    let handle_dir = async |result: &mut ReadGlobResult,
+                            entry_path: RcStr,
+                            segment: &RcStr,
+                            path: &FileSystemPath| {
+        if glob_value.can_match_in_directory(&entry_path) {
+            result.inner.insert(
+                segment.clone(),
+                read_glob_inner(entry_path, path.clone(), glob)
+                    .to_resolved()
+                    .await?,
+            );
+        }
+        anyhow::Ok(())
+    };
+
+    // If there is a parent folder, we need to check if something further up can be matched by the
+    // glob. To prevent an infinite loop where we go up and down repeatedly, only do this if the
+    // prefix has the pattern `../../..`.
+    if let Ok(parent) = directory.join("..")
+        && (prefix.is_empty() || prefix.split("/").all(|s| s == ".."))
+    {
+        let entry_path: RcStr = if prefix.is_empty() {
+            rcstr!("..")
+        } else {
+            format!("{prefix}/..").into()
+        };
+        if glob_value.can_match_in_directory(&entry_path) {
+            result.inner.insert(
+                rcstr!(".."),
+                read_glob_inner(entry_path, parent, glob)
+                    .to_resolved()
+                    .await?,
+            );
+        }
+    }
+
     match &*dir {
         DirectoryContent::Entries(entries) => {
             for (segment, entry) in entries.iter() {
-                // This is redundant with logic inside of `read_dir` but here we track it separately
-                // so we don't follow symlinks.
+                // Ensure that there are no infinite loops, but don't resolve
+                resolve_symlink_safely(entry.clone()).await?;
```

Comment:
Why do we check that here? Isn't that kind of out of scope for this function?

Reply:
Ah I see. This is not checking if the symlink is a cycle, but it's actually checking if the symlink points to a parent directory which would result in a infinite result set for the glob with `**`